core-scalapb/src/main/scala/com/avast/grpc/jsonbridge/scalapb/ScalaPBServiceHandlers.scala

```diff
@@ -0,0 +1,132 @@
+package com.avast.grpc.jsonbridge.scalapb
+
+import java.lang.reflect.{InvocationTargetException, Method}
+
+import cats.effect.Async
+import cats.implicits._
+import com.avast.grpc.jsonbridge.GrpcJsonBridge.GrpcMethodName
+import com.avast.grpc.jsonbridge.{BridgeError, JavaGenericHelper, ReflectionGrpcJsonBridge}
+import com.avast.grpc.jsonbridge.ReflectionGrpcJsonBridge.{HandlerFunc, ServiceHandlers}
+import com.fasterxml.jackson.core.JsonParseException
+import com.typesafe.scalalogging.StrictLogging
+import io.grpc._
+import io.grpc.stub.AbstractStub
+import scalapb.grpc.ConcreteProtoFileDescriptorSupplier
+import scalapb.{GeneratedMessage, GeneratedMessageCompanion}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.{existentials, higherKinds}
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success, Try}
+
+private[jsonbridge] object ScalaPBServiceHandlers extends ServiceHandlers with StrictLogging {
+  def createServiceHandlers[F[+ _]](ec: ExecutionContext)(inProcessChannel: ManagedChannel)(ssd: ServerServiceDefinition)(
+      implicit F: Async[F]): Map[GrpcMethodName, HandlerFunc[F]] = {
+    val futureStubCtor = createFutureStubCtor(ssd.getServiceDescriptor, inProcessChannel)
+    ssd.getMethods.asScala
+      .filter(ReflectionGrpcJsonBridge.isSupportedMethod)
+      .map(createHandler(ec)(futureStubCtor)(_))
+      .toMap
+  }
+
+  private val printer = new scalapb.json4s.Printer().includingDefaultValueFields
+
+  private val parser = new scalapb.json4s.Parser()
+  private val parserMethod = parser.getClass.getDeclaredMethods
+    .find(_.getName == "fromJsonString")
+    .getOrElse(sys.error(s"Method 'fromJsonString' not found on ${parser.getClass}"))
+  private def parse(input: String, companion: GeneratedMessageCompanion[_]): Either[Throwable, GeneratedMessage] =
+    Try(parserMethod.invoke(parser, input, companion).asInstanceOf[GeneratedMessage]) match {
+      case Failure(ie: InvocationTargetException) => Left(ie.getCause)
+      case Failure(e) => Left(e)
+      case Success(m) => Right(m)
+    }
+
+  private def createFutureStubCtor(sd: ServiceDescriptor, inProcessChannel: Channel): () => AbstractStub[_] = {
+    val serviceCompanionClassName = getServiceCompanionClassName(sd)
+    logger.debug(s"Obtaining instance of $serviceCompanionClassName")
+    val serviceCompanionClass = Class.forName(serviceCompanionClassName)
+    val serviceCompanion = serviceCompanionClass.getDeclaredField("MODULE$").get(null)
+    val method = serviceCompanionClass.getDeclaredMethod("stub", classOf[Channel])
+    () =>
```

Comment:
Is there a good reason why this needs to be called later or could you just call it immediately and return the result instead?

Reply:
The reason is that `Class.forName`, `getDeclaredField` and `getDeclaredMethod` would have significant performance impact.

Reply:
I don't get it. These methods are called anyway so they do have the performance impact. The only thing that is postponed is calling `method.invoke` and I don't see the reason for that.

Reply:
The point is that `Class.forName`, `getDeclaredField` and `getDeclaredMethod` are very slow and so it's better if they are called just once (as the class is constructed).

The `invoke` method is called every time (of course), and its performance is OK (maybe could be slightly better with `MethodHandle`).

Check e.g. [this fresh article (in Czech)](https://deadbeef.k47.cz/jvm-reflexe.html).